packages/dev/core/src/Materials/meshDebugPluginMaterial.ts

```diff
@@ -0,0 +1,442 @@
+    /**
+     * 2nd color of checkerboard grid in UV1 or UV2 modes.
+     * Defaults to (0.5, 0.5, 0.5).
+     */
+    uvColor2?: Vector3;
+    /**
+     * Identifying color of this material in MATERIALIDS mode.
+     * Defaults to a randomly-generated color.
+     */
+    materialColor?: Vector3;
+}
+
+/** @internal */
+export class MeshDebugDefines extends MaterialDefines {
+    MODE: MeshDebugMode = MeshDebugMode.NONE;
+    MULTIPLYDEBUG: boolean = true;
+}
+
+/**
+ * Plugin that implements various mesh debug visualizations,
+ * List of available visualizations can be found in MeshDebugMode enum.
+ */
+export class MeshDebugPluginMaterial extends MaterialPluginBase {
+    private _isEnabled: boolean = true;
+    /**
+     * Whether the plugin is enabled or disabled.
+     * Defaults to true.
+     */
+    @serialize()
+    @expandToProperty("_markAllDefinesAsDirty")
+    public isEnabled: boolean = true;
+
+    private _mode: MeshDebugMode = MeshDebugMode.NONE;
+    /**
+     * Current mesh debug mode.
+     * Defaults to NONE.
+     */
+    @serialize()
+    @expandToProperty("_markAllDefinesAsDirty")
+    public mode: MeshDebugMode = MeshDebugMode.NONE;
+
+    /**
+     * Options for the plugin.
+     * See MeshDebugOptions enum for defaults.
+     */
+    private _options: MeshDebugOptions = {};
+
+    /** @internal */
+    public _markAllDefinesAsDirty(): void {
+        this._enable(this._isEnabled);
+        this.markAllDefinesAsDirty();
+    }
+
+    /**
+     * Creates a new MeshDebugPluginMaterial
+     * @param material Material to attach the mesh debug plugin to
+     * @param options Options for the mesh debug plugin
+     */
+    constructor(material: PBRBaseMaterial | StandardMaterial, options?: MeshDebugOptions) {
+        options = options ?? {};
+
+        const defines = new MeshDebugDefines();
+        defines.MODE = options.mode ?? defines.MODE;
+        defines.MULTIPLYDEBUG = options.multiplyDebug ?? defines.MULTIPLYDEBUG;
+        super(material, "MeshDebug", 200, defines);
+
+        options.mode = defines.MODE;
+        options.multiplyDebug = defines.MULTIPLYDEBUG;
+        options.shadedDiffuseColor = options.shadedDiffuseColor ?? new Vector3(1, 1, 1);
+        options.shadedSpecularColor = options.shadedSpecularColor ?? new Vector3(0.8, 0.8, 0.8);
+        options.shadedSpecularPower = options.shadedSpecularPower ?? 10;
+        options.wireframeThickness = options.wireframeThickness ?? 0.7;
+        options.wireframeColor1 = options.wireframeColor1 ?? new Vector3(0, 0, 0);
+        options.wireframeColor2 = options.wireframeColor2 ?? new Vector3(0.9, 0.9, 0.9);
+        options.vertexRadius = options.vertexRadius ?? 1.2;
+        options.uvScale = options.uvScale ?? 20;
+        options.uvColor1 = options.uvColor1 ?? new Vector3(1, 1, 1);
+        options.uvColor2 = options.uvColor2 ?? new Vector3(0.5, 0.5, 0.5);
+        options.materialColor = options.materialColor ?? this._getRandomColor();

```

Comment:
Is it not an issue it changes on every run ?

Reply:
Yes, the random color will change on every run. Should this be changed?

I thought about seeding the random colors with the genRand() function from helperFunctions.fx, but I couldn't think of a great value to seed with from the shader.

(For my own use case, I plan to pass in a fixed materialColor that corresponds to each material's index in the scene so that it's not random)

Reply:
Let s add @PatrickRyanMS to this subject to see how we can do this part ?